{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# os.chdir(\"/scratch/project/tcr_neoantigen/misc/jaz/scripts\")\n",
    "\n",
    "from pathlib import Path\n",
    "from pyfaidx import Fasta\n",
    "from tqdm import tqdm\n",
    "\n",
    "from _utils import (\n",
    "    find,\n",
    "    read_and_filter,\n",
    "    create_result_list,\n",
    "    extract_result,\n",
    "    extract_exon_info,\n",
    "    filter_exon_pos,\n",
    "    generate_windows,\n",
    "    get_sequences_indel,\n",
    "    get_sequences_substitution,\n",
    "    complementary_sequence,\n",
    "    reverse_complement,\n",
    "    flanking_lower_positions,\n",
    "    # print_windows,\n",
    ")\n",
    "\n",
    "# define head paths\n",
    "# HG38FOLDER = Path(\"/scratch/project/tcr_neoantigen/resources/references/hg38\")\n",
    "# INPUTFOLDER = Path(\n",
    "#     \"/scratch/project/tcr_neoantigen/results/cSCC_BC_seq_data_10_patients/nextNEOpi\"\n",
    "# )\n",
    "HG38FOLDER = Path(\"../resources\")\n",
    "INPUTFOLDER = Path(\"../data/nextNEOpi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "FastaNotFoundError",
     "evalue": "Cannot read FASTA from file ../resources/GRCh38.d1.vd1.fa",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pyfaidx/__init__.py:391\u001b[0m, in \u001b[0;36mFaidx.__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, duplicate_action, filt_function, one_based_attributes, read_long_names, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m    390\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 391\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfile \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mr+b\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mmutable\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrb\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    392\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mIOError\u001b[39;00m:\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../resources/GRCh38.d1.vd1.fa'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mFastaNotFoundError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 8\u001b[0m\n\u001b[1;32m      6\u001b[0m fasta_file \u001b[38;5;241m=\u001b[39m HG38FOLDER \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGRCh38.d1.vd1.fa\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      7\u001b[0m gtf_file \u001b[38;5;241m=\u001b[39m HG38FOLDER \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgencode.v33.primary_assembly.annotation.gtf\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 8\u001b[0m refgen \u001b[38;5;241m=\u001b[39m \u001b[43mFasta\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfasta_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      9\u001b[0m exon_info \u001b[38;5;241m=\u001b[39m extract_exon_info(gtf_file)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pyfaidx/__init__.py:1090\u001b[0m, in \u001b[0;36mFasta.__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, filt_function, one_based_attributes, read_long_names, duplicate_action, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m   1084\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1085\u001b[0m \u001b[38;5;124;03mAn object that provides a pygr compatible interface.\u001b[39;00m\n\u001b[1;32m   1086\u001b[0m \u001b[38;5;124;03mfilename:  name of fasta file or fsspec.core.OpenFile instance\u001b[39;00m\n\u001b[1;32m   1087\u001b[0m \u001b[38;5;124;03mindexname: name of index file or fsspec.core.OpenFile instance\u001b[39;00m\n\u001b[1;32m   1088\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1089\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmutable \u001b[38;5;241m=\u001b[39m mutable\n\u001b[0;32m-> 1090\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfaidx \u001b[38;5;241m=\u001b[39m \u001b[43mFaidx\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1091\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1092\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindexname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindexname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1093\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey_function\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkey_function\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1094\u001b[0m \u001b[43m    \u001b[49m\u001b[43mas_raw\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mas_raw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1095\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdefault_seq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdefault_seq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1096\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstrict_bounds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstrict_bounds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1097\u001b[0m \u001b[43m    \u001b[49m\u001b[43mread_ahead\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mread_ahead\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1098\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmutable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmutable\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1099\u001b[0m \u001b[43m    \u001b[49m\u001b[43msplit_char\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msplit_char\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1100\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfilt_function\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfilt_function\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1101\u001b[0m \u001b[43m    \u001b[49m\u001b[43mone_based_attributes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mone_based_attributes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1102\u001b[0m \u001b[43m    \u001b[49m\u001b[43mread_long_names\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mread_long_names\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1103\u001b[0m \u001b[43m    \u001b[49m\u001b[43mduplicate_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mduplicate_action\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1104\u001b[0m \u001b[43m    \u001b[49m\u001b[43msequence_always_upper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msequence_always_upper\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1105\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrebuild\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrebuild\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1106\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbuild_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbuild_index\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1108\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilename \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfaidx\u001b[38;5;241m.\u001b[39mfilename\n\u001b[1;32m   1110\u001b[0m _record_constructor \u001b[38;5;241m=\u001b[39m MutableFastaRecord \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmutable \u001b[38;5;28;01melse\u001b[39;00m FastaRecord\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pyfaidx/__init__.py:393\u001b[0m, in \u001b[0;36mFaidx.__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, duplicate_action, filt_function, one_based_attributes, read_long_names, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m    391\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfile \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(filename, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr+b\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m mutable \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrb\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    392\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mIOError\u001b[39;00m:\n\u001b[0;32m--> 393\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m FastaNotFoundError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot read FASTA from file \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m filename)\n\u001b[1;32m    394\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    396\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mFastaNotFoundError\u001b[0m: Cannot read FASTA from file ../resources/GRCh38.d1.vd1.fa"
     ]
    }
   ],
   "source": [
    "# import reference and data\n",
    "# fasta_file = HG38FOLDER / \"gdc\" / \"GRCh38.d1.vd1\" / \"fasta\" / \"GRCh38.d1.vd1.fa\"\n",
    "# gtf_file = HG38FOLDER / \"annotation\" / \"gencode.v33.primary_assembly.annotation.gtf\"\n",
    "# refgen = Fasta(filename=fasta_file)\n",
    "# exon_info = extract_exon_info(gtf_file)\n",
    "fasta_file = HG38FOLDER / \"GRCh38.d1.vd1.fa\"\n",
    "gtf_file = HG38FOLDER / \"gencode.v33.primary_assembly.annotation.gtf\"\n",
    "refgen = Fasta(filename=fasta_file)\n",
    "exon_info = extract_exon_info(gtf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [\n",
    "    \"2020135\",\n",
    "    \"2020239_WO1\",\n",
    "    \"2020246_NO1\",\n",
    "    \"2020260_WO1\",\n",
    "    \"2020281_WO1\",\n",
    "    \"2021111_MO1\",\n",
    "    \"DES001\",\n",
    "    \"DES002\",\n",
    "    \"DES002_001\",\n",
    "    \"DES002_002\",\n",
    "    \"DES010\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../nextNEOpi/2020135/analyses/2020135/05_vep/tables/high_confidence/2020135_hc_vep.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/Library/CloudStorage/OneDrive-TheUniversityofQueensland/Documents/GitHub/minigene_design/scripts/_utils.py:36\u001b[0m, in \u001b[0;36mread_and_filter\u001b[0;34m(file_input_path, keep_from_lines)\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 36\u001b[0m     df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_input_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     37\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:  \u001b[38;5;66;03m# from Alicia's clean_up_files.py\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    857\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m     \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m     handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m        \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m        \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    863\u001b[0m \u001b[43m        \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    864\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    865\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    866\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m     \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../nextNEOpi/2020135/analyses/2020135/05_vep/tables/high_confidence/2020135_hc_vep.txt'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 13\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m sample \u001b[38;5;129;01min\u001b[39;00m samples:\n\u001b[1;32m      3\u001b[0m     file_input_path \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m      4\u001b[0m         INPUTFOLDER\n\u001b[1;32m      5\u001b[0m         \u001b[38;5;241m/\u001b[39m sample\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[38;5;241m/\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00msample\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_hc_vep.txt\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     12\u001b[0m     )\n\u001b[0;32m---> 13\u001b[0m     results[sample] \u001b[38;5;241m=\u001b[39m \u001b[43mread_and_filter\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_input_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m sample \u001b[38;5;129;01min\u001b[39;00m samples:\n\u001b[1;32m     16\u001b[0m     \u001b[38;5;28mprint\u001b[39m(sample, results[sample]\u001b[38;5;241m.\u001b[39mshape)\n",
      "File \u001b[0;32m~/Library/CloudStorage/OneDrive-TheUniversityofQueensland/Documents/GitHub/minigene_design/scripts/_utils.py:38\u001b[0m, in \u001b[0;36mread_and_filter\u001b[0;34m(file_input_path, keep_from_lines)\u001b[0m\n\u001b[1;32m     36\u001b[0m     df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(file_input_path, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     37\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:  \u001b[38;5;66;03m# from Alicia's clean_up_files.py\u001b[39;00m\n\u001b[0;32m---> 38\u001b[0m     \u001b[43mshutil\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     39\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfile_input_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     40\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfile_input_path\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparent\u001b[49m\n\u001b[1;32m     41\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_input_path\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstem\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m_old\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mfile_input_path\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msuffix\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     42\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     43\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(file_input_path, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m file:\n\u001b[1;32m     44\u001b[0m         lines \u001b[38;5;241m=\u001b[39m file\u001b[38;5;241m.\u001b[39mreadlines()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/shutil.py:419\u001b[0m, in \u001b[0;36mcopy\u001b[0;34m(src, dst, follow_symlinks)\u001b[0m\n\u001b[1;32m    417\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39misdir(dst):\n\u001b[1;32m    418\u001b[0m     dst \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(dst, os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mbasename(src))\n\u001b[0;32m--> 419\u001b[0m \u001b[43mcopyfile\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdst\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfollow_symlinks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_symlinks\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    420\u001b[0m copymode(src, dst, follow_symlinks\u001b[38;5;241m=\u001b[39mfollow_symlinks)\n\u001b[1;32m    421\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m dst\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/dandelion/lib/python3.11/shutil.py:256\u001b[0m, in \u001b[0;36mcopyfile\u001b[0;34m(src, dst, follow_symlinks)\u001b[0m\n\u001b[1;32m    254\u001b[0m     os\u001b[38;5;241m.\u001b[39msymlink(os\u001b[38;5;241m.\u001b[39mreadlink(src), dst)\n\u001b[1;32m    255\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 256\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrb\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m fsrc:\n\u001b[1;32m    257\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    258\u001b[0m             \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(dst, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m fdst:\n\u001b[1;32m    259\u001b[0m                 \u001b[38;5;66;03m# macOS\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../nextNEOpi/2020135/analyses/2020135/05_vep/tables/high_confidence/2020135_hc_vep.txt'"
     ]
    }
   ],
   "source": [
    "results = {}\n",
    "for sample in samples:\n",
    "    file_input_path = (\n",
    "        INPUTFOLDER\n",
    "        / sample\n",
    "        / \"analyses\"\n",
    "        / sample\n",
    "        / \"05_vep\"\n",
    "        / \"tables\"\n",
    "        / \"high_confidence\"\n",
    "        / f\"{sample}_hc_vep.txt\"\n",
    "    )\n",
    "    results[sample] = read_and_filter(file_input_path)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].shape)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].VARIANT_CLASS.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5/5 [00:00<00:00, 1353.96it/s]\n",
      "100%|██████████| 102/102 [00:00<00:00, 2781.95it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 1214.43it/s]\n",
      "100%|██████████| 88/88 [00:00<00:00, 5212.60it/s]\n",
      "100%|██████████| 3/3 [00:00<00:00, 1543.16it/s]\n",
      "100%|██████████| 67/67 [00:00<00:00, 6856.28it/s]\n",
      "100%|██████████| 2/2 [00:00<00:00, 1962.25it/s]\n",
      "100%|██████████| 25/25 [00:00<00:00, 4711.64it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1285.41it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 1334.70it/s]\n",
      "100%|██████████| 24/24 [00:00<00:00, 4792.58it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 595.27it/s]\n",
      "100%|██████████| 31/31 [00:00<00:00, 4649.84it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 2789.23it/s]\n",
      "100%|██████████| 17/17 [00:00<00:00, 2017.80it/s]\n",
      "100%|██████████| 3711/3711 [00:00<00:00, 5047.54it/s]\n",
      "100%|██████████| 432/432 [00:00<00:00, 5877.14it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1628.86it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 1320.83it/s]\n",
      "100%|██████████| 1107/1107 [00:00<00:00, 5289.31it/s]\n",
      "100%|██████████| 122/122 [00:00<00:00, 5088.00it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 3474.50it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 2402.92it/s]\n",
      "100%|██████████| 4913/4913 [00:00<00:00, 6142.65it/s]\n",
      "100%|██████████| 542/542 [00:00<00:00, 7578.29it/s]\n",
      "100%|██████████| 165/165 [00:00<00:00, 5751.88it/s]\n",
      "100%|██████████| 34/34 [00:00<00:00, 6151.86it/s]\n"
     ]
    }
   ],
   "source": [
    "final_results = {}\n",
    "for sample in samples:\n",
    "    df = results[sample].copy()\n",
    "    mut_dict = create_result_list(df)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"insertion\")\n",
    "    final_results[sample] = {}\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_indel(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"deletion\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_indel(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"SNV\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_substitution(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"substitution\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_substitution(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['2020135', '2020239_WO1', '2020246_NO1', '2020260_WO1', '2020281_WO1', '2021111_MO1', 'DES001', 'DES002', 'DES002_001', 'DES002_002', 'DES010'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys([11, 18, 27, 37, 55, 74, 87, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 88, 89, 90, 91, 92, 93, 94])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results[\"2020239_WO1\"].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_list = []\n",
    "for sample in final_results:\n",
    "    for mut in final_results[sample]:\n",
    "        if final_results[sample][mut][\"ref\"] is not None:\n",
    "            for protein in final_results[sample][mut][\"ref\"]:\n",
    "                for frame in final_results[sample][mut][\"ref\"][protein]:\n",
    "                    for i, seq in enumerate(\n",
    "                        final_results[sample][mut][\"ref\"][protein][frame]\n",
    "                    ):\n",
    "                        mut_info = {}\n",
    "                        for key, record in final_results[sample][mut][\n",
    "                            \"mut_info\"\n",
    "                        ].items():\n",
    "                            mut_info[key] = record\n",
    "                        mut_info.update(\n",
    "                            {\n",
    "                                \"minigene\": seq,\n",
    "                                \"minigene_id\": str(sample)\n",
    "                                + \"_\"\n",
    "                                + str(mut)\n",
    "                                + \"_\"\n",
    "                                + protein\n",
    "                                + \"_\"\n",
    "                                + str(frame)\n",
    "                                + \"_\"\n",
    "                                + str(i)\n",
    "                                + \"_ref\",\n",
    "                            }\n",
    "                        )\n",
    "                        ref_list.append(mut_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_1_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GCTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_2_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_3_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>AGCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGC...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_4_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190649</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_1_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190650</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_2_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190651</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGGG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_0_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190652</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_1_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190653</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_2_0_ref</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190654 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       reference variant amino_acid_ref amino_acid_var codon_ref codon_var   \n",
       "0              C       -              T              X       aCc        ac  \\\n",
       "1              C       -              T              X       aCc        ac   \n",
       "2              C       -              T              X       aCc        ac   \n",
       "3              C       -              T              X       aCc        ac   \n",
       "4              C       -              T              X       aCc        ac   \n",
       "...          ...     ...            ...            ...       ...       ...   \n",
       "190649        CC      TT              G              K       GGa       AAa   \n",
       "190650        CC      TT              G              K       GGa       AAa   \n",
       "190651        CC      AT              G              I       GGc       ATc   \n",
       "190652        CC      AT              G              I       GGc       ATc   \n",
       "190653        CC      AT              G              I       GGc       ATc   \n",
       "\n",
       "       chromosome  mutation_location  mutation_location2 variant_class   \n",
       "0            chr2          219483369           219483369      deletion  \\\n",
       "1            chr2          219483369           219483369      deletion   \n",
       "2            chr2          219483369           219483369      deletion   \n",
       "3            chr2          219483369           219483369      deletion   \n",
       "4            chr2          219483369           219483369      deletion   \n",
       "...           ...                ...                 ...           ...   \n",
       "190649       chrX          108221301           108221302  substitution   \n",
       "190650       chrX          108221301           108221302  substitution   \n",
       "190651       chrX          153804140           153804141  substitution   \n",
       "190652       chrX          153804140           153804141  substitution   \n",
       "190653       chrX          153804140           153804141  substitution   \n",
       "\n",
       "                gene_id       protein_id    transcript_id gene_symbol strand   \n",
       "0       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1  \\\n",
       "1       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "2       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "3       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "4       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "...                 ...              ...              ...         ...    ...   \n",
       "190649  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190650  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190651  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190652  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190653  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "\n",
       "                                                 minigene   \n",
       "0       CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...  \\\n",
       "1       CCCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGG...   \n",
       "2       GCTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAG...   \n",
       "3       GGGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTG...   \n",
       "4       AGCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGC...   \n",
       "...                                                   ...   \n",
       "190649  GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGG...   \n",
       "190650  GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...   \n",
       "190651  GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGGG...   \n",
       "190652  GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGG...   \n",
       "190653  CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...   \n",
       "\n",
       "                                 minigene_id  \n",
       "0       2020135_14_ENSP00000311684.7_0_0_ref  \n",
       "1       2020135_14_ENSP00000311684.7_0_1_ref  \n",
       "2       2020135_14_ENSP00000311684.7_0_2_ref  \n",
       "3       2020135_14_ENSP00000311684.7_0_3_ref  \n",
       "4       2020135_14_ENSP00000311684.7_0_4_ref  \n",
       "...                                      ...  \n",
       "190649  DES010_195_ENSP00000334733.7_1_0_ref  \n",
       "190650  DES010_195_ENSP00000334733.7_2_0_ref  \n",
       "190651  DES010_197_ENSP00000164640.4_0_0_ref  \n",
       "190652  DES010_197_ENSP00000164640.4_1_0_ref  \n",
       "190653  DES010_197_ENSP00000164640.4_2_0_ref  \n",
       "\n",
       "[190654 rows x 17 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df = pd.DataFrame(ref_list)\n",
    "ref_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_list = []\n",
    "for sample in final_results:\n",
    "    for mut in final_results[sample]:\n",
    "        if final_results[sample][mut][\"var\"] is not None:\n",
    "            for protein in final_results[sample][mut][\"var\"]:\n",
    "                for frame in final_results[sample][mut][\"var\"][protein]:\n",
    "                    for i, seq in enumerate(\n",
    "                        final_results[sample][mut][\"var\"][protein][frame]\n",
    "                    ):\n",
    "                        mut_info = {}\n",
    "                        for key, record in final_results[sample][mut][\n",
    "                            \"mut_info\"\n",
    "                        ].items():\n",
    "                            mut_info[key] = record\n",
    "                        mut_info.update(\n",
    "                            {\n",
    "                                \"minigene\": seq,\n",
    "                                \"minigene_id\": str(sample)\n",
    "                                + \"_\"\n",
    "                                + str(mut)\n",
    "                                + \"_\"\n",
    "                                + protein\n",
    "                                + \"_\"\n",
    "                                + str(frame)\n",
    "                                + \"_\"\n",
    "                                + str(i)\n",
    "                                + \"_var\",\n",
    "                            }\n",
    "                        )\n",
    "                        var_list.append(mut_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGGA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_1_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAGC...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_2_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTGA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_3_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGCA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_4_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190644</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGA...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_1_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190645</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_2_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190646</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGAT...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_0_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190647</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGA...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_1_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190648</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_2_0_var</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190649 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       reference variant amino_acid_ref amino_acid_var codon_ref codon_var   \n",
       "0              C       -              T              X       aCc        ac  \\\n",
       "1              C       -              T              X       aCc        ac   \n",
       "2              C       -              T              X       aCc        ac   \n",
       "3              C       -              T              X       aCc        ac   \n",
       "4              C       -              T              X       aCc        ac   \n",
       "...          ...     ...            ...            ...       ...       ...   \n",
       "190644        CC      TT              G              K       GGa       AAa   \n",
       "190645        CC      TT              G              K       GGa       AAa   \n",
       "190646        CC      AT              G              I       GGc       ATc   \n",
       "190647        CC      AT              G              I       GGc       ATc   \n",
       "190648        CC      AT              G              I       GGc       ATc   \n",
       "\n",
       "       chromosome  mutation_location  mutation_location2 variant_class   \n",
       "0            chr2          219483369           219483369      deletion  \\\n",
       "1            chr2          219483369           219483369      deletion   \n",
       "2            chr2          219483369           219483369      deletion   \n",
       "3            chr2          219483369           219483369      deletion   \n",
       "4            chr2          219483369           219483369      deletion   \n",
       "...           ...                ...                 ...           ...   \n",
       "190644       chrX          108221301           108221302  substitution   \n",
       "190645       chrX          108221301           108221302  substitution   \n",
       "190646       chrX          153804140           153804141  substitution   \n",
       "190647       chrX          153804140           153804141  substitution   \n",
       "190648       chrX          153804140           153804141  substitution   \n",
       "\n",
       "                gene_id       protein_id    transcript_id gene_symbol strand   \n",
       "0       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1  \\\n",
       "1       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "2       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "3       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "4       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "...                 ...              ...              ...         ...    ...   \n",
       "190644  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190645  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190646  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190647  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190648  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "\n",
       "                                                 minigene   \n",
       "0       CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...  \\\n",
       "1       CCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGGA...   \n",
       "2       CTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAGC...   \n",
       "3       GGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTGA...   \n",
       "4       GCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGCA...   \n",
       "...                                                   ...   \n",
       "190644  GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGA...   \n",
       "190645  GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...   \n",
       "190646  GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGAT...   \n",
       "190647  GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGA...   \n",
       "190648  CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...   \n",
       "\n",
       "                                 minigene_id  \n",
       "0       2020135_14_ENSP00000311684.7_0_0_var  \n",
       "1       2020135_14_ENSP00000311684.7_0_1_var  \n",
       "2       2020135_14_ENSP00000311684.7_0_2_var  \n",
       "3       2020135_14_ENSP00000311684.7_0_3_var  \n",
       "4       2020135_14_ENSP00000311684.7_0_4_var  \n",
       "...                                      ...  \n",
       "190644  DES010_195_ENSP00000334733.7_1_0_var  \n",
       "190645  DES010_195_ENSP00000334733.7_2_0_var  \n",
       "190646  DES010_197_ENSP00000164640.4_0_0_var  \n",
       "190647  DES010_197_ENSP00000164640.4_1_0_var  \n",
       "190648  DES010_197_ENSP00000164640.4_2_0_var  \n",
       "\n",
       "[190649 rows x 17 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_df = pd.DataFrame(var_list)\n",
    "var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df.to_csv(\"final_minigenes_control.csv\", index=False)\n",
    "var_df.to_csv(\"final_minigenes_variant.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
