{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# os.chdir(\"/scratch/project/tcr_neoantigen/misc/jaz/scripts\")\n",
    "\n",
    "from pathlib import Path\n",
    "from pyfaidx import Fasta\n",
    "from tqdm import tqdm\n",
    "\n",
    "from _utils import (\n",
    "    find,\n",
    "    read_and_filter,\n",
    "    create_result_list,\n",
    "    extract_result,\n",
    "    extract_exon_info,\n",
    "    filter_exon_pos,\n",
    "    generate_windows,\n",
    "    get_sequences_indel,\n",
    "    get_sequences_substitution,\n",
    "    complementary_sequence,\n",
    "    reverse_complement,\n",
    "    flanking_lower_positions,\n",
    "    # print_windows,\n",
    ")\n",
    "\n",
    "# define head paths\n",
    "# HG38FOLDER = Path(\"/scratch/project/tcr_neoantigen/resources/references/hg38\")\n",
    "# INPUTFOLDER = Path(\n",
    "#     \"/scratch/project/tcr_neoantigen/results/cSCC_BC_seq_data_10_patients/nextNEOpi\"\n",
    "# )\n",
    "HG38FOLDER = Path(\"../resources\")\n",
    "INPUTFOLDER = Path(\"../data/nextNEOpi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import reference and data\n",
    "# fasta_file = HG38FOLDER / \"gdc\" / \"GRCh38.d1.vd1\" / \"fasta\" / \"GRCh38.d1.vd1.fa\"\n",
    "# gtf_file = HG38FOLDER / \"annotation\" / \"gencode.v33.primary_assembly.annotation.gtf\"\n",
    "# refgen = Fasta(filename=fasta_file)\n",
    "# exon_info = extract_exon_info(gtf_file)\n",
    "fasta_file = HG38FOLDER / \"GRCh38.d1.vd1.fa\"\n",
    "gtf_file = HG38FOLDER / \"gencode.v33.primary_assembly.annotation.gtf\"\n",
    "refgen = Fasta(filename=fasta_file)\n",
    "exon_info = extract_exon_info(gtf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [\n",
    "    \"2020135\",\n",
    "    \"2020239_WO1\",\n",
    "    \"2020246_NO1\",\n",
    "    \"2020260_WO1\",\n",
    "    \"2020281_WO1\",\n",
    "    \"2021111_MO1\",\n",
    "    \"DES001\",\n",
    "    \"DES002\",\n",
    "    \"DES002_001\",\n",
    "    \"DES002_002\",\n",
    "    \"DES010\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2020135 (107, 91)\n",
      "2020239_WO1 (95, 91)\n",
      "2020246_NO1 (70, 91)\n",
      "2020260_WO1 (27, 91)\n",
      "2020281_WO1 (29, 91)\n",
      "2021111_MO1 (32, 91)\n",
      "DES001 (4164, 91)\n",
      "DES002 (1240, 91)\n",
      "DES002_001 (5481, 91)\n",
      "DES002_002 (0, 91)\n",
      "DES010 (199, 91)\n",
      "2020135 ['SNV' 'deletion']\n",
      "2020239_WO1 ['SNV' 'deletion']\n",
      "2020246_NO1 ['SNV' 'deletion']\n",
      "2020260_WO1 ['SNV' 'deletion']\n",
      "2020281_WO1 ['SNV' 'deletion' 'insertion']\n",
      "2021111_MO1 ['SNV' 'insertion']\n",
      "DES001 ['SNV' 'substitution' 'deletion' 'insertion']\n",
      "DES002 ['SNV' 'substitution' 'deletion' 'insertion']\n",
      "DES002_001 ['SNV' 'substitution' 'deletion' 'insertion']\n",
      "DES002_002 []\n",
      "DES010 ['SNV' 'substitution']\n"
     ]
    }
   ],
   "source": [
    "results = {}\n",
    "for sample in samples:\n",
    "    file_input_path = (\n",
    "        INPUTFOLDER\n",
    "        / sample\n",
    "        / \"analyses\"\n",
    "        / sample\n",
    "        / \"05_vep\"\n",
    "        / \"tables\"\n",
    "        / \"high_confidence\"\n",
    "        / f\"{sample}_hc_vep.txt\"\n",
    "    )\n",
    "    results[sample] = read_and_filter(file_input_path)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].shape)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].VARIANT_CLASS.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5/5 [00:00<00:00, 1174.02it/s]\n",
      "100%|██████████| 102/102 [00:00<00:00, 1649.59it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 890.56it/s]\n",
      "100%|██████████| 88/88 [00:00<00:00, 3604.48it/s]\n",
      "100%|██████████| 3/3 [00:00<00:00, 1770.00it/s]\n",
      "100%|██████████| 67/67 [00:00<00:00, 4508.20it/s]\n",
      "100%|██████████| 2/2 [00:00<00:00, 1216.80it/s]\n",
      "100%|██████████| 25/25 [00:00<00:00, 3517.30it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 702.33it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 760.46it/s]\n",
      "100%|██████████| 24/24 [00:00<00:00, 3659.69it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 520.51it/s]\n",
      "100%|██████████| 31/31 [00:00<00:00, 3669.56it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 2200.58it/s]\n",
      "100%|██████████| 17/17 [00:00<00:00, 1381.76it/s]\n",
      "100%|██████████| 3711/3711 [00:00<00:00, 3961.31it/s]\n",
      "100%|██████████| 432/432 [00:00<00:00, 4715.18it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1620.67it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 1131.21it/s]\n",
      "100%|██████████| 1107/1107 [00:00<00:00, 4398.94it/s]\n",
      "100%|██████████| 122/122 [00:00<00:00, 4339.02it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 2529.74it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 1909.50it/s]\n",
      "100%|██████████| 4913/4913 [00:00<00:00, 5289.25it/s]\n",
      "100%|██████████| 542/542 [00:00<00:00, 6566.21it/s]\n",
      "100%|██████████| 165/165 [00:00<00:00, 4848.67it/s]\n",
      "100%|██████████| 34/34 [00:00<00:00, 4884.45it/s]\n"
     ]
    }
   ],
   "source": [
    "final_results = {}\n",
    "for sample in samples:\n",
    "    df = results[sample].copy()\n",
    "    mut_dict = create_result_list(df)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"insertion\")\n",
    "    final_results[sample] = {}\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_indel(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"deletion\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_indel(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"SNV\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_substitution(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"substitution\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_substitution(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['2020135', '2020239_WO1', '2020246_NO1', '2020260_WO1', '2020281_WO1', '2021111_MO1', 'DES001', 'DES002', 'DES002_001', 'DES002_002', 'DES010'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys([11, 18, 27, 37, 55, 74, 87, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 88, 89, 90, 91, 92, 93, 94])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results[\"2020239_WO1\"].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_list = []\n",
    "for sample in final_results:\n",
    "    for mut in final_results[sample]:\n",
    "        if final_results[sample][mut][\"ref\"] is not None:\n",
    "            for protein in final_results[sample][mut][\"ref\"]:\n",
    "                for frame in final_results[sample][mut][\"ref\"][protein]:\n",
    "                    for i, seq in enumerate(\n",
    "                        final_results[sample][mut][\"ref\"][protein][frame]\n",
    "                    ):\n",
    "                        mut_info = {}\n",
    "                        for key, record in final_results[sample][mut][\n",
    "                            \"mut_info\"\n",
    "                        ].items():\n",
    "                            mut_info[key] = record\n",
    "                        mut_info.update(\n",
    "                            {\n",
    "                                \"minigene\": seq,\n",
    "                                \"minigene_id\": str(sample)\n",
    "                                + \"_\"\n",
    "                                + str(mut)\n",
    "                                + \"_\"\n",
    "                                + protein\n",
    "                                + \"_\"\n",
    "                                + str(frame)\n",
    "                                + \"_\"\n",
    "                                + str(i)\n",
    "                                + \"_ref\",\n",
    "                            }\n",
    "                        )\n",
    "                        ref_list.append(mut_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_1_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GCTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_2_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_3_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>AGCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGC...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_4_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190649</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_1_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190650</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_2_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190651</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGGG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_0_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190652</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_1_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190653</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_2_0_ref</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190654 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       reference variant amino_acid_ref amino_acid_var codon_ref codon_var   \n",
       "0              C       -              T              X       aCc        ac  \\\n",
       "1              C       -              T              X       aCc        ac   \n",
       "2              C       -              T              X       aCc        ac   \n",
       "3              C       -              T              X       aCc        ac   \n",
       "4              C       -              T              X       aCc        ac   \n",
       "...          ...     ...            ...            ...       ...       ...   \n",
       "190649        CC      TT              G              K       GGa       AAa   \n",
       "190650        CC      TT              G              K       GGa       AAa   \n",
       "190651        CC      AT              G              I       GGc       ATc   \n",
       "190652        CC      AT              G              I       GGc       ATc   \n",
       "190653        CC      AT              G              I       GGc       ATc   \n",
       "\n",
       "       chromosome  mutation_location  mutation_location2 variant_class   \n",
       "0            chr2          219483369           219483369      deletion  \\\n",
       "1            chr2          219483369           219483369      deletion   \n",
       "2            chr2          219483369           219483369      deletion   \n",
       "3            chr2          219483369           219483369      deletion   \n",
       "4            chr2          219483369           219483369      deletion   \n",
       "...           ...                ...                 ...           ...   \n",
       "190649       chrX          108221301           108221302  substitution   \n",
       "190650       chrX          108221301           108221302  substitution   \n",
       "190651       chrX          153804140           153804141  substitution   \n",
       "190652       chrX          153804140           153804141  substitution   \n",
       "190653       chrX          153804140           153804141  substitution   \n",
       "\n",
       "                gene_id       protein_id    transcript_id gene_symbol strand   \n",
       "0       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1  \\\n",
       "1       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "2       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "3       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "4       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "...                 ...              ...              ...         ...    ...   \n",
       "190649  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190650  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190651  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190652  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190653  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "\n",
       "                                                 minigene   \n",
       "0       CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...  \\\n",
       "1       CCCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGG...   \n",
       "2       GCTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAG...   \n",
       "3       GGGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTG...   \n",
       "4       AGCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGC...   \n",
       "...                                                   ...   \n",
       "190649  GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGG...   \n",
       "190650  GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...   \n",
       "190651  GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGGG...   \n",
       "190652  GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGG...   \n",
       "190653  CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...   \n",
       "\n",
       "                                 minigene_id  \n",
       "0       2020135_14_ENSP00000311684.7_0_0_ref  \n",
       "1       2020135_14_ENSP00000311684.7_0_1_ref  \n",
       "2       2020135_14_ENSP00000311684.7_0_2_ref  \n",
       "3       2020135_14_ENSP00000311684.7_0_3_ref  \n",
       "4       2020135_14_ENSP00000311684.7_0_4_ref  \n",
       "...                                      ...  \n",
       "190649  DES010_195_ENSP00000334733.7_1_0_ref  \n",
       "190650  DES010_195_ENSP00000334733.7_2_0_ref  \n",
       "190651  DES010_197_ENSP00000164640.4_0_0_ref  \n",
       "190652  DES010_197_ENSP00000164640.4_1_0_ref  \n",
       "190653  DES010_197_ENSP00000164640.4_2_0_ref  \n",
       "\n",
       "[190654 rows x 17 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df = pd.DataFrame(ref_list)\n",
    "ref_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_list = []\n",
    "for sample in final_results:\n",
    "    for mut in final_results[sample]:\n",
    "        if final_results[sample][mut][\"var\"] is not None:\n",
    "            for protein in final_results[sample][mut][\"var\"]:\n",
    "                for frame in final_results[sample][mut][\"var\"][protein]:\n",
    "                    for i, seq in enumerate(\n",
    "                        final_results[sample][mut][\"var\"][protein][frame]\n",
    "                    ):\n",
    "                        mut_info = {}\n",
    "                        for key, record in final_results[sample][mut][\n",
    "                            \"mut_info\"\n",
    "                        ].items():\n",
    "                            mut_info[key] = record\n",
    "                        mut_info.update(\n",
    "                            {\n",
    "                                \"minigene\": seq,\n",
    "                                \"minigene_id\": str(sample)\n",
    "                                + \"_\"\n",
    "                                + str(mut)\n",
    "                                + \"_\"\n",
    "                                + protein\n",
    "                                + \"_\"\n",
    "                                + str(frame)\n",
    "                                + \"_\"\n",
    "                                + str(i)\n",
    "                                + \"_var\",\n",
    "                            }\n",
    "                        )\n",
    "                        var_list.append(mut_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGGA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_1_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAGC...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_2_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTGA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_3_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGCA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_4_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190644</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGA...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_1_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190645</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_2_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190646</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGAT...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_0_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190647</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGA...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_1_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190648</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_2_0_var</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190649 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       reference variant amino_acid_ref amino_acid_var codon_ref codon_var   \n",
       "0              C       -              T              X       aCc        ac  \\\n",
       "1              C       -              T              X       aCc        ac   \n",
       "2              C       -              T              X       aCc        ac   \n",
       "3              C       -              T              X       aCc        ac   \n",
       "4              C       -              T              X       aCc        ac   \n",
       "...          ...     ...            ...            ...       ...       ...   \n",
       "190644        CC      TT              G              K       GGa       AAa   \n",
       "190645        CC      TT              G              K       GGa       AAa   \n",
       "190646        CC      AT              G              I       GGc       ATc   \n",
       "190647        CC      AT              G              I       GGc       ATc   \n",
       "190648        CC      AT              G              I       GGc       ATc   \n",
       "\n",
       "       chromosome  mutation_location  mutation_location2 variant_class   \n",
       "0            chr2          219483369           219483369      deletion  \\\n",
       "1            chr2          219483369           219483369      deletion   \n",
       "2            chr2          219483369           219483369      deletion   \n",
       "3            chr2          219483369           219483369      deletion   \n",
       "4            chr2          219483369           219483369      deletion   \n",
       "...           ...                ...                 ...           ...   \n",
       "190644       chrX          108221301           108221302  substitution   \n",
       "190645       chrX          108221301           108221302  substitution   \n",
       "190646       chrX          153804140           153804141  substitution   \n",
       "190647       chrX          153804140           153804141  substitution   \n",
       "190648       chrX          153804140           153804141  substitution   \n",
       "\n",
       "                gene_id       protein_id    transcript_id gene_symbol strand   \n",
       "0       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1  \\\n",
       "1       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "2       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "3       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "4       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG      1   \n",
       "...                 ...              ...              ...         ...    ...   \n",
       "190644  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190645  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6     -1   \n",
       "190646  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190647  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "190648  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4     -1   \n",
       "\n",
       "                                                 minigene   \n",
       "0       CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...  \\\n",
       "1       CCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGGA...   \n",
       "2       CTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAGC...   \n",
       "3       GGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTGA...   \n",
       "4       GCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGCA...   \n",
       "...                                                   ...   \n",
       "190644  GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGA...   \n",
       "190645  GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...   \n",
       "190646  GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGAT...   \n",
       "190647  GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGA...   \n",
       "190648  CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...   \n",
       "\n",
       "                                 minigene_id  \n",
       "0       2020135_14_ENSP00000311684.7_0_0_var  \n",
       "1       2020135_14_ENSP00000311684.7_0_1_var  \n",
       "2       2020135_14_ENSP00000311684.7_0_2_var  \n",
       "3       2020135_14_ENSP00000311684.7_0_3_var  \n",
       "4       2020135_14_ENSP00000311684.7_0_4_var  \n",
       "...                                      ...  \n",
       "190644  DES010_195_ENSP00000334733.7_1_0_var  \n",
       "190645  DES010_195_ENSP00000334733.7_2_0_var  \n",
       "190646  DES010_197_ENSP00000164640.4_0_0_var  \n",
       "190647  DES010_197_ENSP00000164640.4_1_0_var  \n",
       "190648  DES010_197_ENSP00000164640.4_2_0_var  \n",
       "\n",
       "[190649 rows x 17 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_df = pd.DataFrame(var_list)\n",
    "var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df.to_csv(\"final_minigenes_control.csv\", index=False)\n",
    "var_df.to_csv(\"final_minigenes_variant.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
