{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# os.chdir(\"/scratch/project/tcr_neoantigen/misc/jaz/scripts\")\n",
    "\n",
    "from pathlib import Path\n",
    "from pyfaidx import Fasta\n",
    "from tqdm import tqdm\n",
    "\n",
    "from _utils import (\n",
    "    find,\n",
    "    read_and_filter,\n",
    "    create_result_list,\n",
    "    extract_result,\n",
    "    extract_exon_info,\n",
    "    filter_exon_pos,\n",
    "    generate_windows,\n",
    "    get_sequences_indel,\n",
    "    get_sequences_substitution,\n",
    "    complementary_sequence,\n",
    "    reverse_complement,\n",
    "    flanking_lower_positions,\n",
    "    # print_windows,\n",
    ")\n",
    "\n",
    "# define head paths\n",
    "# HG38FOLDER = Path(\"/scratch/project/tcr_neoantigen/resources/references/hg38\")\n",
    "# INPUTFOLDER = Path(\n",
    "#     \"/scratch/project/tcr_neoantigen/results/cSCC_BC_seq_data_10_patients/nextNEOpi\"\n",
    "# )\n",
    "HG38FOLDER = Path(\"../resources\")\n",
    "INPUTFOLDER = Path(\"../data/nextNEOpi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import reference and data\n",
    "fasta_file = HG38FOLDER / \"gdc\" / \"GRCh38.d1.vd1\" / \"fasta\" / \"GRCh38.d1.vd1.fa\"\n",
    "gtf_file = HG38FOLDER / \"annotation\" / \"gencode.v33.primary_assembly.annotation.gtf\"\n",
    "# refgen = Fasta(filename=fasta_file)\n",
    "# exon_info = extract_exon_info(gtf_file)\n",
    "# fasta_file = HG38FOLDER / \"GRCh38.d1.vd1.fa\"\n",
    "# gtf_file = HG38FOLDER / \"gencode.v33.primary_assembly.annotation.gtf\"\n",
    "refgen = Fasta(filename=fasta_file)\n",
    "exon_info = extract_exon_info(gtf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [\n",
    "    \"2020135\",\n",
    "    \"2020239_WO1\",\n",
    "    \"2020246_NO1\",\n",
    "    \"2020260_WO1\",\n",
    "    \"2020281_WO1\",\n",
    "    \"2021111_MO1\",\n",
    "    \"DES001\",\n",
    "    \"DES002\",\n",
    "    \"DES002_001\",\n",
    "    \"DES002_002\",\n",
    "    \"DES010\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test (3, 91)\n",
      "test ['substitution' 'SNV' 'insertion']\n"
     ]
    }
   ],
   "source": [
    "results = {}\n",
    "samples = [\"test\"]\n",
    "for sample in samples:\n",
    "    # file_input_path = (\n",
    "    #     INPUTFOLDER\n",
    "    #     / sample\n",
    "    #     / \"analyses\"\n",
    "    #     / sample\n",
    "    #     / \"05_vep\"\n",
    "    #     / \"tables\"\n",
    "    #     / \"high_confidence\"\n",
    "    #     / f\"{sample}_hc_vep.txt\"\n",
    "    # )\n",
    "    file_input_path = f\"../test/{sample}_hc_vep.txt\"\n",
    "    results[sample] = read_and_filter(file_input_path)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].shape)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].VARIANT_CLASS.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00, 999.83it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 500.10it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1001.51it/s]\n"
     ]
    }
   ],
   "source": [
    "final_results = {}\n",
    "for sample in samples:\n",
    "    df = results[sample].copy()\n",
    "    mut_dict = create_result_list(df)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"insertion\")\n",
    "    final_results[sample] = {}\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_indel(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"deletion\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_indel(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"SNV\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_substitution(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"substitution\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            final_results[sample][mut] = {}\n",
    "            final_results[sample][mut][\"ref\"], final_results[sample][mut][\"var\"] = (\n",
    "                get_sequences_substitution(\n",
    "                    mut_info=extract_result(mut_dict, mut),\n",
    "                    exon_info=exon_info,\n",
    "                    fasta=refgen,\n",
    "                )\n",
    "            )\n",
    "            final_results[sample][mut][\"mut_info\"] = extract_result(mut_dict, mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['test'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys([2, 1, 0])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results[\"test\"].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_list = []\n",
    "for sample in final_results:\n",
    "    for mut in final_results[sample]:\n",
    "        if final_results[sample][mut][\"ref\"] is not None:\n",
    "            for protein in final_results[sample][mut][\"ref\"]:\n",
    "                for frame in final_results[sample][mut][\"ref\"][protein]:\n",
    "                    for i, seq in enumerate(\n",
    "                        final_results[sample][mut][\"ref\"][protein][frame]\n",
    "                    ):\n",
    "                        mut_info = {}\n",
    "                        for key, record in final_results[sample][mut][\n",
    "                            \"mut_info\"\n",
    "                        ].items():\n",
    "                            mut_info[key] = record\n",
    "                        mut_info.update(\n",
    "                            {\n",
    "                                \"minigene\": seq,\n",
    "                                \"minigene_id\": str(sample)\n",
    "                                + \"_\"\n",
    "                                + str(mut)\n",
    "                                + \"_\"\n",
    "                                + protein\n",
    "                                + \"_\"\n",
    "                                + str(frame)\n",
    "                                + \"_\"\n",
    "                                + str(i)\n",
    "                                + \"_ref\",\n",
    "                            }\n",
    "                        )\n",
    "                        ref_list.append(mut_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>AATTAGGAAGGCCATGGAATCTGCTGAACAAAAGGAACAAGGTTTA...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_0_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>AAACCTTGTTCCTTTTGTTCAGCAGATTCCATGGCCTTCCTAATTT...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_1_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>AACTGGATCTGAGCTTGTTTCTTATCATTCAACATTTGCCTGTGAT...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_2_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCAAGGCTCTTAACTGCTCTTCACTGAAATAACCCGGTAGTTGTT...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_3_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>TACTGAGTTTTTTTTGTCGCTGCTAACTGTATGTTAGCTCTTTCAG...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_4_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGCCCTTTCTCTGCCCCATCCCTACCCTAGCCTTGCTCTCAGCCA...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_31_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGATAGTCACTGGGCTCCCTGTGACTTCTGACCCTGACACCCCTC...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_32_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGGACTCTGCCTGGGCTGGAGTCTAGGGCTGGGGCTACATTTGGC...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_33_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>CTGTACTGGCTGAGGACAGGGGAGGGAGTGAAGTTGGTTTGGGGTG...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_34_ref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>GTTTGGGGTGGCCTGTGTTGCCACTCTCAGCACCCCACATTTGCAT...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_35_ref</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference variant amino_acid_ref amino_acid_var codon_ref codon_var  \\\n",
       "0           -       G              S             SX       tca      tcGa   \n",
       "1           -       G              S             SX       tca      tcGa   \n",
       "2           -       G              S             SX       tca      tcGa   \n",
       "3           -       G              S             SX       tca      tcGa   \n",
       "4           -       G              S             SX       tca      tcGa   \n",
       "..        ...     ...            ...            ...       ...       ...   \n",
       "300        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "301        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "302        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "303        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "304        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "\n",
       "    chromosome  mutation_location  mutation_location2 variant_class  \\\n",
       "0        chr13           32379450            32379451     insertion   \n",
       "1        chr13           32379450            32379451     insertion   \n",
       "2        chr13           32379450            32379451     insertion   \n",
       "3        chr13           32379450            32379451     insertion   \n",
       "4        chr13           32379450            32379451     insertion   \n",
       "..         ...                ...                 ...           ...   \n",
       "300       chr1            1355129             1355130  substitution   \n",
       "301       chr1            1355129             1355130  substitution   \n",
       "302       chr1            1355129             1355130  substitution   \n",
       "303       chr1            1355129             1355130  substitution   \n",
       "304       chr1            1355129             1355130  substitution   \n",
       "\n",
       "             gene_id       protein_id    transcript_id gene_symbol  strand  \\\n",
       "0    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "1    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "2    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "3    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "4    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "..               ...              ...              ...         ...     ...   \n",
       "300  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "301  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "302  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "303  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "304  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "\n",
       "                                              minigene  \\\n",
       "0    AATTAGGAAGGCCATGGAATCTGCTGAACAAAAGGAACAAGGTTTA...   \n",
       "1    AAACCTTGTTCCTTTTGTTCAGCAGATTCCATGGCCTTCCTAATTT...   \n",
       "2    AACTGGATCTGAGCTTGTTTCTTATCATTCAACATTTGCCTGTGAT...   \n",
       "3    TTCAAGGCTCTTAACTGCTCTTCACTGAAATAACCCGGTAGTTGTT...   \n",
       "4    TACTGAGTTTTTTTTGTCGCTGCTAACTGTATGTTAGCTCTTTCAG...   \n",
       "..                                                 ...   \n",
       "300  TTGCCCTTTCTCTGCCCCATCCCTACCCTAGCCTTGCTCTCAGCCA...   \n",
       "301  TTGATAGTCACTGGGCTCCCTGTGACTTCTGACCCTGACACCCCTC...   \n",
       "302  TTGGACTCTGCCTGGGCTGGAGTCTAGGGCTGGGGCTACATTTGGC...   \n",
       "303  CTGTACTGGCTGAGGACAGGGGAGGGAGTGAAGTTGGTTTGGGGTG...   \n",
       "304  GTTTGGGGTGGCCTGTGTTGCCACTCTCAGCACCCCACATTTGCAT...   \n",
       "\n",
       "                           minigene_id  \n",
       "0     test_2_ENSP00000369497.3_0_0_ref  \n",
       "1     test_2_ENSP00000369497.3_0_1_ref  \n",
       "2     test_2_ENSP00000369497.3_0_2_ref  \n",
       "3     test_2_ENSP00000369497.3_0_3_ref  \n",
       "4     test_2_ENSP00000369497.3_0_4_ref  \n",
       "..                                 ...  \n",
       "300  test_0_ENSP00000307887.6_2_31_ref  \n",
       "301  test_0_ENSP00000307887.6_2_32_ref  \n",
       "302  test_0_ENSP00000307887.6_2_33_ref  \n",
       "303  test_0_ENSP00000307887.6_2_34_ref  \n",
       "304  test_0_ENSP00000307887.6_2_35_ref  \n",
       "\n",
       "[305 rows x 17 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df = pd.DataFrame(ref_list)\n",
    "ref_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_list = []\n",
    "for sample in final_results:\n",
    "    for mut in final_results[sample]:\n",
    "        if final_results[sample][mut][\"var\"] is not None:\n",
    "            for protein in final_results[sample][mut][\"var\"]:\n",
    "                for frame in final_results[sample][mut][\"var\"][protein]:\n",
    "                    for i, seq in enumerate(\n",
    "                        final_results[sample][mut][\"var\"][protein][frame]\n",
    "                    ):\n",
    "                        mut_info = {}\n",
    "                        for key, record in final_results[sample][mut][\n",
    "                            \"mut_info\"\n",
    "                        ].items():\n",
    "                            mut_info[key] = record\n",
    "                        mut_info.update(\n",
    "                            {\n",
    "                                \"minigene\": seq,\n",
    "                                \"minigene_id\": str(sample)\n",
    "                                + \"_\"\n",
    "                                + str(mut)\n",
    "                                + \"_\"\n",
    "                                + protein\n",
    "                                + \"_\"\n",
    "                                + str(frame)\n",
    "                                + \"_\"\n",
    "                                + str(i)\n",
    "                                + \"_var\",\n",
    "                            }\n",
    "                        )\n",
    "                        var_list.append(mut_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>ATTAGGAAGGCCATGGAATCTGCTGAACAAAAGGAACAAGGTTTAT...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_0_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>TAAACCTTGTTCCTTTTGTTCAGCAGATTCCATGGCCTTCCTAATT...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_1_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>CAACTGGATCTGAGCTTGTTTCTTATCATTCAACATTTGCCTGTGA...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_2_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>ATTCAAGGCTCTTAACTGCTCTTCACTGAAATAACCCGGTAGTTGT...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_3_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>SX</td>\n",
       "      <td>tca</td>\n",
       "      <td>tcGa</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32379450</td>\n",
       "      <td>32379451</td>\n",
       "      <td>insertion</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1</td>\n",
       "      <td>ATACTGAGTTTTTTTTGTCGCTGCTAACTGTATGTTAGCTCTTTCA...</td>\n",
       "      <td>test_2_ENSP00000369497.3_0_4_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGCCCTTTCTCTGCCCCATCCCTACCCTAGCCTTGCTCTCAGCCA...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_31_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGATAGTCACTGGGCTCCCTGTGACTTCTGACCCTGACACCCCTC...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_32_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGGACTCTGCCTGGGCTGGAGTCTAGGGCTGGGGCTACATTTGGC...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_33_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>CTGTACTGGCTGAGGACAGGGGAGGGAGTGAAGTTGGTTTGGGGTG...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_34_var</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>WD</td>\n",
       "      <td>*N</td>\n",
       "      <td>tgGGac</td>\n",
       "      <td>tgAAac</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1355129</td>\n",
       "      <td>1355130</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENSP00000307887</td>\n",
       "      <td>ENST00000309212</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>-1</td>\n",
       "      <td>GTTTGGGGTGGCCTGTGTTGCCACTCTCAGCACCCCACATTTGCAT...</td>\n",
       "      <td>test_0_ENSP00000307887.6_2_35_var</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference variant amino_acid_ref amino_acid_var codon_ref codon_var  \\\n",
       "0           -       G              S             SX       tca      tcGa   \n",
       "1           -       G              S             SX       tca      tcGa   \n",
       "2           -       G              S             SX       tca      tcGa   \n",
       "3           -       G              S             SX       tca      tcGa   \n",
       "4           -       G              S             SX       tca      tcGa   \n",
       "..        ...     ...            ...            ...       ...       ...   \n",
       "300        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "301        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "302        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "303        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "304        CC      TT             WD             *N    tgGGac    tgAAac   \n",
       "\n",
       "    chromosome  mutation_location  mutation_location2 variant_class  \\\n",
       "0        chr13           32379450            32379451     insertion   \n",
       "1        chr13           32379450            32379451     insertion   \n",
       "2        chr13           32379450            32379451     insertion   \n",
       "3        chr13           32379450            32379451     insertion   \n",
       "4        chr13           32379450            32379451     insertion   \n",
       "..         ...                ...                 ...           ...   \n",
       "300       chr1            1355129             1355130  substitution   \n",
       "301       chr1            1355129             1355130  substitution   \n",
       "302       chr1            1355129             1355130  substitution   \n",
       "303       chr1            1355129             1355130  substitution   \n",
       "304       chr1            1355129             1355130  substitution   \n",
       "\n",
       "             gene_id       protein_id    transcript_id gene_symbol  strand  \\\n",
       "0    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "1    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "2    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "3    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "4    ENSG00000139618  ENSP00000369497  ENST00000380152       BRCA2       1   \n",
       "..               ...              ...              ...         ...     ...   \n",
       "300  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "301  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "302  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "303  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "304  ENSG00000162576  ENSP00000307887  ENST00000309212       MXRA8      -1   \n",
       "\n",
       "                                              minigene  \\\n",
       "0    ATTAGGAAGGCCATGGAATCTGCTGAACAAAAGGAACAAGGTTTAT...   \n",
       "1    TAAACCTTGTTCCTTTTGTTCAGCAGATTCCATGGCCTTCCTAATT...   \n",
       "2    CAACTGGATCTGAGCTTGTTTCTTATCATTCAACATTTGCCTGTGA...   \n",
       "3    ATTCAAGGCTCTTAACTGCTCTTCACTGAAATAACCCGGTAGTTGT...   \n",
       "4    ATACTGAGTTTTTTTTGTCGCTGCTAACTGTATGTTAGCTCTTTCA...   \n",
       "..                                                 ...   \n",
       "300  TTGCCCTTTCTCTGCCCCATCCCTACCCTAGCCTTGCTCTCAGCCA...   \n",
       "301  TTGATAGTCACTGGGCTCCCTGTGACTTCTGACCCTGACACCCCTC...   \n",
       "302  TTGGACTCTGCCTGGGCTGGAGTCTAGGGCTGGGGCTACATTTGGC...   \n",
       "303  CTGTACTGGCTGAGGACAGGGGAGGGAGTGAAGTTGGTTTGGGGTG...   \n",
       "304  GTTTGGGGTGGCCTGTGTTGCCACTCTCAGCACCCCACATTTGCAT...   \n",
       "\n",
       "                           minigene_id  \n",
       "0     test_2_ENSP00000369497.3_0_0_var  \n",
       "1     test_2_ENSP00000369497.3_0_1_var  \n",
       "2     test_2_ENSP00000369497.3_0_2_var  \n",
       "3     test_2_ENSP00000369497.3_0_3_var  \n",
       "4     test_2_ENSP00000369497.3_0_4_var  \n",
       "..                                 ...  \n",
       "300  test_0_ENSP00000307887.6_2_31_var  \n",
       "301  test_0_ENSP00000307887.6_2_32_var  \n",
       "302  test_0_ENSP00000307887.6_2_33_var  \n",
       "303  test_0_ENSP00000307887.6_2_34_var  \n",
       "304  test_0_ENSP00000307887.6_2_35_var  \n",
       "\n",
       "[305 rows x 17 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_df = pd.DataFrame(var_list)\n",
    "var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df.to_csv(\"final_minigenes_control.csv\", index=False)\n",
    "var_df.to_csv(\"final_minigenes_variant.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
