{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# os.chdir(\"/scratch/project/tcr_neoantigen/misc/jaz/scripts\")\n",
    "\n",
    "from pathlib import Path\n",
    "from pyfaidx import Fasta\n",
    "from tqdm import tqdm\n",
    "\n",
    "from _utils import (\n",
    "    find,\n",
    "    read_and_filter,\n",
    "    create_result_list,\n",
    "    extract_result,\n",
    "    extract_exon_info,\n",
    "    filter_exon_pos,\n",
    "    get_sequences_indel,\n",
    "    get_sequences_substitution,\n",
    "    complementary_sequence,\n",
    "    reverse_complement,\n",
    "    flanking_lower_positions,\n",
    "    # print_windows,\n",
    ")\n",
    "\n",
    "# define head paths\n",
    "HG38FOLDER = Path(\"/Users/Ali/Desktop/Long Covid/minigene_design/Reference/GRCh38.d1.vd1/fasta\")\n",
    "INPUTFOLDER = Path(\"/Users/Ali/Desktop/Long Covid/minigene_design/input\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import reference and data\n",
    "fasta_file = HG38FOLDER / \"GRCh38.d1.vd1.fa\"\n",
    "gtf_file = HG38FOLDER / \"gencode.v33.primary_assembly.annotation.gtf\"\n",
    "\n",
    "# Ensure the path exists\n",
    "if not fasta_file.exists():\n",
    "    raise FileNotFoundError(f\"The specified FASTA file was not found: {fasta_file}\")\n",
    "\n",
    "if not gtf_file.exists():\n",
    "    raise FileNotFoundError(f\"The specified GTF file was not found: {gtf_file}\")\n",
    "\n",
    "refgen = Fasta(filename=fasta_file)\n",
    "exon_info = extract_exon_info(gtf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [\n",
    "    \"2020135\",\n",
    "    \"2020239_WO1\",\n",
    "    \"2020246_NO1\",\n",
    "    \"2020260_WO1\",\n",
    "    \"2020281_WO1\",\n",
    "    \"2021111_MO1\",\n",
    "    \"DES001\",\n",
    "    \"DES002\",\n",
    "    \"DES002_001\",\n",
    "    \"DES002_002\",\n",
    "    \"DES010\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Users/Ali/Desktop/Long Covid/minigene_design/input/2020135/analyses/2020135/05_vep/tables/high_confidence/2020135_hc_vep.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/Desktop/Long Covid/minigene_design/scripts/_utils.py:33\u001b[0m, in \u001b[0;36mread_and_filter\u001b[0;34m(file_input_path, keep_from_lines)\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 33\u001b[0m     df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mread_csv(file_input_path, sep\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\\t\u001b[39;49;00m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m     34\u001b[0m \u001b[39mexcept\u001b[39;00m:  \u001b[39m# from Alicia's clean_up_files.py\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/minigenes/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1024\u001b[0m kwds\u001b[39m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[39mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n",
      "File \u001b[0;32m~/anaconda3/envs/minigenes/lib/python3.12/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    619\u001b[0m \u001b[39m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[39m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwds)\n\u001b[1;32m    622\u001b[0m \u001b[39mif\u001b[39;00m chunksize \u001b[39mor\u001b[39;00m iterator:\n",
      "File \u001b[0;32m~/anaconda3/envs/minigenes/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1619\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles: IOHandles \u001b[39m|\u001b[39m \u001b[39mNone\u001b[39;00m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_engine(f, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mengine)\n",
      "File \u001b[0;32m~/anaconda3/envs/minigenes/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1880\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1879\u001b[0m         mode \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m-> 1880\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39m=\u001b[39m get_handle(\n\u001b[1;32m   1881\u001b[0m     f,\n\u001b[1;32m   1882\u001b[0m     mode,\n\u001b[1;32m   1883\u001b[0m     encoding\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1884\u001b[0m     compression\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mcompression\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1885\u001b[0m     memory_map\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mmemory_map\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mFalse\u001b[39;49;00m),\n\u001b[1;32m   1886\u001b[0m     is_text\u001b[39m=\u001b[39;49mis_text,\n\u001b[1;32m   1887\u001b[0m     errors\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding_errors\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mstrict\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[1;32m   1888\u001b[0m     storage_options\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mstorage_options\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1889\u001b[0m )\n\u001b[1;32m   1890\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/minigenes/lib/python3.12/site-packages/pandas/io/common.py:873\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    871\u001b[0m \u001b[39mif\u001b[39;00m ioargs\u001b[39m.\u001b[39mencoding \u001b[39mand\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m ioargs\u001b[39m.\u001b[39mmode:\n\u001b[1;32m    872\u001b[0m     \u001b[39m# Encoding\u001b[39;00m\n\u001b[0;32m--> 873\u001b[0m     handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39;49m(\n\u001b[1;32m    874\u001b[0m         handle,\n\u001b[1;32m    875\u001b[0m         ioargs\u001b[39m.\u001b[39;49mmode,\n\u001b[1;32m    876\u001b[0m         encoding\u001b[39m=\u001b[39;49mioargs\u001b[39m.\u001b[39;49mencoding,\n\u001b[1;32m    877\u001b[0m         errors\u001b[39m=\u001b[39;49merrors,\n\u001b[1;32m    878\u001b[0m         newline\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    879\u001b[0m     )\n\u001b[1;32m    880\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    881\u001b[0m     \u001b[39m# Binary mode\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/Ali/Desktop/Long Covid/minigene_design/input/2020135/analyses/2020135/05_vep/tables/high_confidence/2020135_hc_vep.txt'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 13\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[39mfor\u001b[39;00m sample \u001b[39min\u001b[39;00m samples:\n\u001b[1;32m      3\u001b[0m     file_input_path \u001b[39m=\u001b[39m (\n\u001b[1;32m      4\u001b[0m         INPUTFOLDER\n\u001b[1;32m      5\u001b[0m         \u001b[39m/\u001b[39m sample\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[39m/\u001b[39m \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00msample\u001b[39m}\u001b[39;00m\u001b[39m_hc_vep.txt\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     12\u001b[0m     )\n\u001b[0;32m---> 13\u001b[0m     results[sample] \u001b[39m=\u001b[39m read_and_filter(file_input_path)\n\u001b[1;32m     15\u001b[0m \u001b[39mfor\u001b[39;00m sample \u001b[39min\u001b[39;00m samples:\n\u001b[1;32m     16\u001b[0m     \u001b[39mprint\u001b[39m(sample, results[sample]\u001b[39m.\u001b[39mshape)\n",
      "File \u001b[0;32m~/Desktop/Long Covid/minigene_design/scripts/_utils.py:35\u001b[0m, in \u001b[0;36mread_and_filter\u001b[0;34m(file_input_path, keep_from_lines)\u001b[0m\n\u001b[1;32m     33\u001b[0m     df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(file_input_path, sep\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     34\u001b[0m \u001b[39mexcept\u001b[39;00m:  \u001b[39m# from Alicia's clean_up_files.py\u001b[39;00m\n\u001b[0;32m---> 35\u001b[0m     shutil\u001b[39m.\u001b[39;49mcopy(\n\u001b[1;32m     36\u001b[0m         file_input_path,\n\u001b[1;32m     37\u001b[0m         file_input_path\u001b[39m.\u001b[39;49mparent \u001b[39m/\u001b[39;49m (file_input_path\u001b[39m.\u001b[39;49mstem \u001b[39m+\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39m_old\u001b[39;49m\u001b[39m\"\u001b[39;49m \u001b[39m+\u001b[39;49m file_input_path\u001b[39m.\u001b[39;49msuffix),\n\u001b[1;32m     38\u001b[0m     )\n\u001b[1;32m     39\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(file_input_path, \u001b[39m\"\u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mas\u001b[39;00m file:\n\u001b[1;32m     40\u001b[0m         lines \u001b[39m=\u001b[39m file\u001b[39m.\u001b[39mreadlines()\n",
      "File \u001b[0;32m~/anaconda3/envs/minigenes/lib/python3.12/shutil.py:435\u001b[0m, in \u001b[0;36mcopy\u001b[0;34m(src, dst, follow_symlinks)\u001b[0m\n\u001b[1;32m    433\u001b[0m \u001b[39mif\u001b[39;00m os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39misdir(dst):\n\u001b[1;32m    434\u001b[0m     dst \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mjoin(dst, os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mbasename(src))\n\u001b[0;32m--> 435\u001b[0m copyfile(src, dst, follow_symlinks\u001b[39m=\u001b[39;49mfollow_symlinks)\n\u001b[1;32m    436\u001b[0m copymode(src, dst, follow_symlinks\u001b[39m=\u001b[39mfollow_symlinks)\n\u001b[1;32m    437\u001b[0m \u001b[39mreturn\u001b[39;00m dst\n",
      "File \u001b[0;32m~/anaconda3/envs/minigenes/lib/python3.12/shutil.py:260\u001b[0m, in \u001b[0;36mcopyfile\u001b[0;34m(src, dst, follow_symlinks)\u001b[0m\n\u001b[1;32m    258\u001b[0m     os\u001b[39m.\u001b[39msymlink(os\u001b[39m.\u001b[39mreadlink(src), dst)\n\u001b[1;32m    259\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 260\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39;49m(src, \u001b[39m'\u001b[39;49m\u001b[39mrb\u001b[39;49m\u001b[39m'\u001b[39;49m) \u001b[39mas\u001b[39;00m fsrc:\n\u001b[1;32m    261\u001b[0m         \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    262\u001b[0m             \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(dst, \u001b[39m'\u001b[39m\u001b[39mwb\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mas\u001b[39;00m fdst:\n\u001b[1;32m    263\u001b[0m                 \u001b[39m# macOS\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/Ali/Desktop/Long Covid/minigene_design/input/2020135/analyses/2020135/05_vep/tables/high_confidence/2020135_hc_vep.txt'"
     ]
    }
   ],
   "source": [
    "results = {}\n",
    "for sample in samples:\n",
    "    file_input_path = (\n",
    "        INPUTFOLDER\n",
    "        / sample\n",
    "        / \"analyses\"\n",
    "        / sample\n",
    "        / \"05_vep\"\n",
    "        / \"tables\"\n",
    "        / \"high_confidence\"\n",
    "        / f\"{sample}_hc_vep.txt\"\n",
    "    )\n",
    "    results[sample] = read_and_filter(file_input_path)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].shape)\n",
    "\n",
    "for sample in samples:\n",
    "    print(sample, results[sample].VARIANT_CLASS.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5/5 [00:00<00:00, 1646.37it/s]\n",
      "100%|██████████| 102/102 [00:00<00:00, 2337.83it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 970.45it/s]\n",
      "100%|██████████| 88/88 [00:00<00:00, 3856.23it/s]\n",
      "100%|██████████| 3/3 [00:00<00:00, 1848.53it/s]\n",
      "100%|██████████| 67/67 [00:00<00:00, 4835.39it/s]\n",
      "100%|██████████| 2/2 [00:00<00:00, 1430.53it/s]\n",
      "100%|██████████| 25/25 [00:00<00:00, 3808.71it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1326.47it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 1059.03it/s]\n",
      "100%|██████████| 24/24 [00:00<00:00, 3078.48it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 604.19it/s]\n",
      "100%|██████████| 31/31 [00:00<00:00, 3525.01it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 731.26it/s]\n",
      "100%|██████████| 17/17 [00:00<00:00, 835.38it/s]\n",
      "100%|██████████| 3711/3711 [00:00<00:00, 4249.83it/s]\n",
      "100%|██████████| 432/432 [00:00<00:00, 5268.17it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1392.99it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 1009.39it/s]\n",
      "100%|██████████| 1107/1107 [00:00<00:00, 4044.25it/s]\n",
      "100%|██████████| 122/122 [00:00<00:00, 4588.87it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 2436.19it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 1756.85it/s]\n",
      "100%|██████████| 4913/4913 [00:01<00:00, 4904.48it/s]\n",
      "100%|██████████| 542/542 [00:00<00:00, 6892.77it/s]\n",
      "100%|██████████| 165/165 [00:00<00:00, 4628.45it/s]\n",
      "100%|██████████| 34/34 [00:00<00:00, 5537.68it/s]\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    df = results[sample].copy()\n",
    "    mut_dict = create_result_list(df)\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"insertion\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            ref_seqs, var_seqs = get_sequences_indel(\n",
    "                mut_info=extract_result(mut_dict, mut),\n",
    "                exon_info=exon_info,\n",
    "                fasta=refgen,\n",
    "            )\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"deletion\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            ref_seqs, var_seqs = get_sequences_indel(\n",
    "                mut_info=extract_result(mut_dict, mut),\n",
    "                exon_info=exon_info,\n",
    "                fasta=refgen,\n",
    "            )\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"SNV\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            ref_seqs, var_seqs = get_sequences_substitution(\n",
    "                mut_info=extract_result(mut_dict, mut),\n",
    "                exon_info=exon_info,\n",
    "                fasta=refgen,\n",
    "            )\n",
    "    mutations = find(mut_dict[\"variant_class\"], \"substitution\")\n",
    "    if len(mutations) > 0:\n",
    "        for mut in tqdm(mutations):\n",
    "            ref_seqs, var_seqs = get_sequences_substitution(\n",
    "                mut_info=extract_result(mut_dict, mut),\n",
    "                exon_info=exon_info,\n",
    "                fasta=refgen,\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = \"TTCGCTCAAGGCTGTGGGTCCACCACCCCAAACCCCTCCACGAAGACACAGGGGCCTGCAGGCTGCCCGGCCAGCGGAGCCCACCCTACCCAGT\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# os.chdir(\"/scratch/project/tcr_neoantigen/misc/jaz/scripts\")\n",
    "\n",
    "from pathlib import Path\n",
    "from pyfaidx import Fasta\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "varient_csv = Path(\"/Users/Ali/Desktop/Long Covid/minigene_design/input/final_minigenes_variant_v2.csv\")\n",
    "control_csv = Path(\"/Users/Ali/Desktop/Long Covid/minigene_design/input/final_minigenes_control_v2.csv\")\n",
    "df1 = pd.read_csv(varient_csv)\n",
    "df2 = pd.read_csv(control_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# situation 1: GAA GAC, CTT CTG, GTC TTC, GTC TTC, CAG AAG\n",
    "dict1={\"GAA\" : \"GAG\",\n",
    "\"CTT\" : \"CTA\",\n",
    "\"GTC\" : \"GTA\",\n",
    "\"CAG\" : \"CAA\"}\n",
    "# situation 2: _GA AGAC, _CT TCT G, _GT CTT C, _CA GAA G\n",
    "dict2={\n",
    "\"AGA\" : \"AGG\",\n",
    "\"GGA\" : \"GGG\",\n",
    "\"CGA\" : \"CGG\",\n",
    "\"TGA\" : \"TAG\",\n",
    "\"ACT\" : \"ACC\",\n",
    "\"GCT\" : \"GCC\",\n",
    "\"CCT\" : \"CCC\",\n",
    "\"TCT\" : \"TCC\",\n",
    "\"AGT\" : \"AGC\",\n",
    "\"GGT\" : \"GGC\",\n",
    "\"CGT\" : \"CGC\",\n",
    "\"TGT\" : \"TGC\",\n",
    "\"ACA\" : \"ACC\",\n",
    "\"GCA\" : \"GCC\",\n",
    "\"CCA\" : \"CCC\",\n",
    "\"TCA\" : \"TCC\"}\n",
    "\n",
    "# situation 3: __G AAG AC, __C TTC TG, __G TCT TC, __C AGA AG\n",
    "dict3={\"AAG\" : \"AAA\",\n",
    "\"ACG\" : \"ACT\",\n",
    "\"AGG\" : \"AGA\",\n",
    "\"TAG\" : \"TGA\",\n",
    "\"TTG\" : \"TTA\",\n",
    "\"TCG\" : \"TCT\",\n",
    "\"CAG\" : \"CAA\",\n",
    "\"CTG\" : \"CTA\",\n",
    "\"CGG\" : \"CGA\",\n",
    "\"GAG\" : \"GAA\",\n",
    "\"GTG\" : \"GTA\",\n",
    "\"GCG\" : \"GCT\",\n",
    "\"GGG\" : \"GGA\",\n",
    "\"AAC\" : \"AAT\",\n",
    "\"ACC\" : \"ACT\",\n",
    "\"AGC\" : \"AGT\",\n",
    "\"ATC\" : \"ATT\",\n",
    "\"TAC\" : \"TAT\",\n",
    "\"TTC\" : \"TTT\",\n",
    "\"TCC\" : \"TCT\",\n",
    "\"TGC\" : \"TGT\",\n",
    "\"CAC\" : \"CAT\",\n",
    "\"CTC\" : \"CTT\",\n",
    "\"CGC\" : \"CGT\",\n",
    "\"CCC\" : \"CCT\",\n",
    "\"GAC\" : \"GAT\",\n",
    "\"GTC\" : \"GTT\",\n",
    "\"GCC\" : \"GCT\",\n",
    "\"GGC\" : \"GGT\"}\n",
    "\n",
    "# situation 4: ATG, TGG\n",
    "dict4={\"AAG\" : \"AAA\",\n",
    "\"TCT\" : \"TCC\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'TTC'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 25\u001b[0m\n\u001b[1;32m     22\u001b[0m \t\t\t\u001b[39melse\u001b[39;00m:\n\u001b[1;32m     23\u001b[0m \t\t\t\t\u001b[39m# becomes situation 2\u001b[39;00m\n\u001b[1;32m     24\u001b[0m \t\t\t\tpos \u001b[39m=\u001b[39m position \u001b[39m+\u001b[39m \u001b[39m1\u001b[39m\n\u001b[0;32m---> 25\u001b[0m \t\t\t\tnew_seq \u001b[39m=\u001b[39m seq[:pos] \u001b[39m+\u001b[39m dict4[seq[pos:pos\u001b[39m+\u001b[39;49m\u001b[39m3\u001b[39;49m]] \u001b[39m+\u001b[39m seq[pos\u001b[39m+\u001b[39m\u001b[39m3\u001b[39m:]\n\u001b[1;32m     27\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m     28\u001b[0m \tnew_seq \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mN/A\u001b[39m\u001b[39m\"\u001b[39m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'TTC'"
     ]
    }
   ],
   "source": [
    "import re\n",
    "bbs1_varient = []\n",
    "for seq in df1[\"minigene\"]:\n",
    "\tfor x in [\"GAAGAC\",\"CTTCTG\",\"GTCTTC\",\"CAGAAG\",]:\n",
    "\t\tif re.search(x, seq):\n",
    "\t\t\t# first need to find the location number of the sequence and see if that number is divisible by 3\n",
    "\t\t\tposition = seq.find(x)\n",
    "\t\t\tif position % 3 == 0:\n",
    "\t\t\t\t# situation 1\n",
    "\t\t\t\tpos = position\n",
    "\t\t\t\tnew_seq = seq[:pos] + dict1[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\telse:\n",
    "\t\t\t\tif (position - 1) % 3 == 0:\n",
    "\t\t\t\t\t# situation 2\n",
    "\t\t\t\t\tpos = position - 1\n",
    "\t\t\t\t\tnew_seq = seq[:pos] + dict2[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\t# situation 3\n",
    "\t\t\t\t\tpos = position - 2\n",
    "\t\t\t\t\tif seq[pos:pos+3] in dict2:\n",
    "\t\t\t\t\t\tnew_seq = seq[:pos] + dict3[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t# becomes situation 2\n",
    "\t\t\t\t\t\tpos = position + 1\n",
    "\t\t\t\t\t\tnew_seq = seq[:pos] + dict4[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\t\t\t\t\t\t\n",
    "\t\telse:\n",
    "\t\t\tnew_seq = \"N/A\"\n",
    "\tbbs1_varient.append(new_seq)\n",
    "\n",
    "\n",
    "df1[\"modified_minigene\"] = bbs1_varient\n",
    "# Save the DataFrame to a CSV file\n",
    "output_csv = Path(\"/Users/Ali/Desktop/Long Covid/minigene_design/output/modified_minigenes_variant_v2.csv\")\n",
    "df1.to_csv(output_csv, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "      <th>modified_minigene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_0_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGGA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_1_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAGC...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_2_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTGA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_3_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGCA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_4_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190644</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGA...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_1_0_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190645</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_2_0_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190646</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGAT...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_0_0_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190647</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGA...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_1_0_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190648</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_2_0_var</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190649 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       reference variant amino_acid_ref amino_acid_var codon_ref codon_var  \\\n",
       "0              C       -              T              X       aCc        ac   \n",
       "1              C       -              T              X       aCc        ac   \n",
       "2              C       -              T              X       aCc        ac   \n",
       "3              C       -              T              X       aCc        ac   \n",
       "4              C       -              T              X       aCc        ac   \n",
       "...          ...     ...            ...            ...       ...       ...   \n",
       "190644        CC      TT              G              K       GGa       AAa   \n",
       "190645        CC      TT              G              K       GGa       AAa   \n",
       "190646        CC      AT              G              I       GGc       ATc   \n",
       "190647        CC      AT              G              I       GGc       ATc   \n",
       "190648        CC      AT              G              I       GGc       ATc   \n",
       "\n",
       "       chromosome  mutation_location  mutation_location2 variant_class  \\\n",
       "0            chr2          219483369           219483369      deletion   \n",
       "1            chr2          219483369           219483369      deletion   \n",
       "2            chr2          219483369           219483369      deletion   \n",
       "3            chr2          219483369           219483369      deletion   \n",
       "4            chr2          219483369           219483369      deletion   \n",
       "...           ...                ...                 ...           ...   \n",
       "190644       chrX          108221301           108221302  substitution   \n",
       "190645       chrX          108221301           108221302  substitution   \n",
       "190646       chrX          153804140           153804141  substitution   \n",
       "190647       chrX          153804140           153804141  substitution   \n",
       "190648       chrX          153804140           153804141  substitution   \n",
       "\n",
       "                gene_id       protein_id    transcript_id gene_symbol  strand  \\\n",
       "0       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "1       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "2       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "3       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "4       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "...                 ...              ...              ...         ...     ...   \n",
       "190644  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6      -1   \n",
       "190645  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6      -1   \n",
       "190646  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4      -1   \n",
       "190647  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4      -1   \n",
       "190648  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4      -1   \n",
       "\n",
       "                                                 minigene  \\\n",
       "0       CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...   \n",
       "1       CCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGGA...   \n",
       "2       CTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAGC...   \n",
       "3       GGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTGA...   \n",
       "4       GCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGCA...   \n",
       "...                                                   ...   \n",
       "190644  GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGA...   \n",
       "190645  GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...   \n",
       "190646  GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGAT...   \n",
       "190647  GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGA...   \n",
       "190648  CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...   \n",
       "\n",
       "                                 minigene_id modified_minigene  \n",
       "0       2020135_14_ENSP00000311684.7_0_0_var               N/A  \n",
       "1       2020135_14_ENSP00000311684.7_0_1_var               N/A  \n",
       "2       2020135_14_ENSP00000311684.7_0_2_var               N/A  \n",
       "3       2020135_14_ENSP00000311684.7_0_3_var               N/A  \n",
       "4       2020135_14_ENSP00000311684.7_0_4_var               N/A  \n",
       "...                                      ...               ...  \n",
       "190644  DES010_195_ENSP00000334733.7_1_0_var               N/A  \n",
       "190645  DES010_195_ENSP00000334733.7_2_0_var               N/A  \n",
       "190646  DES010_197_ENSP00000164640.4_0_0_var               N/A  \n",
       "190647  DES010_197_ENSP00000164640.4_1_0_var               N/A  \n",
       "190648  DES010_197_ENSP00000164640.4_2_0_var               N/A  \n",
       "\n",
       "[190649 rows x 18 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "bbs1_varient = []\n",
    "for seq in df2[\"minigene\"]:\n",
    "\tfor x in [\"GAAGAC\",\"CTTCTG\",\"GTCTTC\",\"CAGAAG\",]:\n",
    "\t\tif re.search(x, seq):\n",
    "\t\t\t# first need to find the location number of the sequence and see if that number is divisible by 3\n",
    "\t\t\tposition = seq.find(x)\n",
    "\t\t\tif position % 3 == 0:\n",
    "\t\t\t\t# situation 1\n",
    "\t\t\t\tpos = position\n",
    "\t\t\t\tnew_seq = seq[:pos] + dict1[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\t\t# print(\"1\",new_seq )\n",
    "\t\t\telse:\n",
    "\t\t\t\tif (position - 1) % 3 == 0:\n",
    "\t\t\t\t\t# situation 2\n",
    "\t\t\t\t\tpos = position - 1\n",
    "\t\t\t\t\tnew_seq = seq[:pos] + dict2[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\t\t\t# print(\"2\",new_seq )\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\t# situation 3\n",
    "\t\t\t\t\tpos = position - 2\n",
    "\t\t\t\t\tif seq[pos:pos+3] in dict2:\n",
    "\t\t\t\t\t\tnew_seq = seq[:pos] + dict3[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\t\t\t\t# print(\"3\",new_seq )\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t# becomes situation 2\n",
    "\t\t\t\t\t\tpos = position - 1\n",
    "\t\t\t\t\t\tnew_seq = seq[:pos] + dict2[seq[pos:pos+3]] + seq[pos+3:]\n",
    "\t\t\t\t\t\t# print(\"4\",new_seq )\t\t\n",
    "\t\telse:\n",
    "\t\t\tnew_seq = \"N/A\"\n",
    "\tbbs1_varient.append(new_seq)\n",
    "\n",
    "\n",
    "df2[\"modified_minigene\"] = bbs1_varient\n",
    "# Save the DataFrame to a CSV file\n",
    "output_csv = Path(\"/Users/Ali/Desktop/Long Covid/minigene_design/output/final_minigenes_control_v2.csv\")\n",
    "df2.to_csv(output_csv, index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>amino_acid_ref</th>\n",
       "      <th>amino_acid_var</th>\n",
       "      <th>codon_ref</th>\n",
       "      <th>codon_var</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mutation_location</th>\n",
       "      <th>mutation_location2</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>strand</th>\n",
       "      <th>minigene</th>\n",
       "      <th>minigene_id</th>\n",
       "      <th>modified_minigene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_0_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_1_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GCTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_2_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTG...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_3_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>aCc</td>\n",
       "      <td>ac</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219483369</td>\n",
       "      <td>219483369</td>\n",
       "      <td>deletion</td>\n",
       "      <td>ENSG00000072195</td>\n",
       "      <td>ENSP00000311684</td>\n",
       "      <td>ENST00000312358</td>\n",
       "      <td>SPEG</td>\n",
       "      <td>1</td>\n",
       "      <td>AGCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGC...</td>\n",
       "      <td>2020135_14_ENSP00000311684.7_0_4_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190649</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_1_0_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190650</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>GGa</td>\n",
       "      <td>AAa</td>\n",
       "      <td>chrX</td>\n",
       "      <td>108221301</td>\n",
       "      <td>108221302</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000197565</td>\n",
       "      <td>ENSP00000334733</td>\n",
       "      <td>ENST00000334504</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...</td>\n",
       "      <td>DES010_195_ENSP00000334733.7_2_0_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190651</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGGG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_0_0_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190652</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_1_0_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190653</th>\n",
       "      <td>CC</td>\n",
       "      <td>AT</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>GGc</td>\n",
       "      <td>ATc</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153804140</td>\n",
       "      <td>153804141</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ENSG00000067840</td>\n",
       "      <td>ENSP00000164640</td>\n",
       "      <td>ENST00000164640</td>\n",
       "      <td>PDZD4</td>\n",
       "      <td>-1</td>\n",
       "      <td>CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...</td>\n",
       "      <td>DES010_197_ENSP00000164640.4_2_0_ref</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190654 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       reference variant amino_acid_ref amino_acid_var codon_ref codon_var  \\\n",
       "0              C       -              T              X       aCc        ac   \n",
       "1              C       -              T              X       aCc        ac   \n",
       "2              C       -              T              X       aCc        ac   \n",
       "3              C       -              T              X       aCc        ac   \n",
       "4              C       -              T              X       aCc        ac   \n",
       "...          ...     ...            ...            ...       ...       ...   \n",
       "190649        CC      TT              G              K       GGa       AAa   \n",
       "190650        CC      TT              G              K       GGa       AAa   \n",
       "190651        CC      AT              G              I       GGc       ATc   \n",
       "190652        CC      AT              G              I       GGc       ATc   \n",
       "190653        CC      AT              G              I       GGc       ATc   \n",
       "\n",
       "       chromosome  mutation_location  mutation_location2 variant_class  \\\n",
       "0            chr2          219483369           219483369      deletion   \n",
       "1            chr2          219483369           219483369      deletion   \n",
       "2            chr2          219483369           219483369      deletion   \n",
       "3            chr2          219483369           219483369      deletion   \n",
       "4            chr2          219483369           219483369      deletion   \n",
       "...           ...                ...                 ...           ...   \n",
       "190649       chrX          108221301           108221302  substitution   \n",
       "190650       chrX          108221301           108221302  substitution   \n",
       "190651       chrX          153804140           153804141  substitution   \n",
       "190652       chrX          153804140           153804141  substitution   \n",
       "190653       chrX          153804140           153804141  substitution   \n",
       "\n",
       "                gene_id       protein_id    transcript_id gene_symbol  strand  \\\n",
       "0       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "1       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "2       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "3       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "4       ENSG00000072195  ENSP00000311684  ENST00000312358        SPEG       1   \n",
       "...                 ...              ...              ...         ...     ...   \n",
       "190649  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6      -1   \n",
       "190650  ENSG00000197565  ENSP00000334733  ENST00000334504      COL4A6      -1   \n",
       "190651  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4      -1   \n",
       "190652  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4      -1   \n",
       "190653  ENSG00000067840  ENSP00000164640  ENST00000164640       PDZD4      -1   \n",
       "\n",
       "                                                 minigene  \\\n",
       "0       CCACTGAGGATGAGGCCCTGGGGACCCCAGAGACTGGGGCTGCCAA...   \n",
       "1       CCCATGGACTGGCAGGAGCAGGGAAGGGCTCCCTCTCAGGACCAGG...   \n",
       "2       GCTCCCAGCCCAGAGGCCCTCCCCTCCCCAGGCCAGGAGCCCGCAG...   \n",
       "3       GGGGCTAGCCCCAGGCGGGGAGAGCTCCGCAGGGGCAGCTCGGCTG...   \n",
       "4       AGCGCCCTGCCCCGGGCCGGGCCGCGGGAGCTGGGCCGGGGCCTGC...   \n",
       "...                                                   ...   \n",
       "190649  GCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGGG...   \n",
       "190650  GGCCCAACAGGTCCTCAAGGATTCACTGGCTCTACTGGTTTATCGG...   \n",
       "190651  GCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGGG...   \n",
       "190652  GGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCGG...   \n",
       "190653  CGGCGGGCCATGGCCGGCAACTCCAACTTGAACCGGACCCCTCCCG...   \n",
       "\n",
       "                                 minigene_id modified_minigene  \n",
       "0       2020135_14_ENSP00000311684.7_0_0_ref               N/A  \n",
       "1       2020135_14_ENSP00000311684.7_0_1_ref               N/A  \n",
       "2       2020135_14_ENSP00000311684.7_0_2_ref               N/A  \n",
       "3       2020135_14_ENSP00000311684.7_0_3_ref               N/A  \n",
       "4       2020135_14_ENSP00000311684.7_0_4_ref               N/A  \n",
       "...                                      ...               ...  \n",
       "190649  DES010_195_ENSP00000334733.7_1_0_ref               N/A  \n",
       "190650  DES010_195_ENSP00000334733.7_2_0_ref               N/A  \n",
       "190651  DES010_197_ENSP00000164640.4_0_0_ref               N/A  \n",
       "190652  DES010_197_ENSP00000164640.4_1_0_ref               N/A  \n",
       "190653  DES010_197_ENSP00000164640.4_2_0_ref               N/A  \n",
       "\n",
       "[190654 rows x 18 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
